src/index/ivf_raft/ivf_raft.cuh

```diff
@@ -158,15 +174,15 @@ class RaftIvfIndexNode : public IndexNode {
                     return metric.error();
                 }
                 if (metric.value() != raft::distance::DistanceType::L2Expanded &&
-                    metric.value() != raft::distance::DistanceType::L2Unexpanded &&
                     metric.value() != raft::distance::DistanceType::InnerProduct) {
                     LOG_KNOWHERE_WARNING_ << "selected metric not supported in RAFT IVF indexes: "
                                           << ivf_raft_cfg.metric_type;
                     return Status::invalid_metric_type;
                 }
-
+                devs_.insert(devs_.begin(), ivf_raft_cfg.gpu_ids.begin(), ivf_raft_cfg.gpu_ids.end());
                 auto scoped_device = detail::device_setter{*ivf_raft_cfg.gpu_ids.begin()};
-                res_ = std::make_unique<raft::device_resources>();
+                auto res_ = std::make_unique<raft::device_resources>(rmm::cuda_stream_per_thread, nullptr,
```

Comment:
Any reason we recreate the `device_resources` with every call? I'm assuming this is to ensure that we are getting the per-thread default stream each time.

I think we might be better served with a stream pool, which can be added directly to the `device_resources`. If we need to ensure that `Train` and `Add` complete before we perform a search, we can synchronize on the whole pool after those calls. Then, we can just pull streams from that pool to perform overlapping searches.

Reply:
Just to add to @wphicks, `device_resources` manages all of the resources that you need to interact w/ the GPU (and with other GPU-accelerated libraries like cublas) and they can be very expensive to create. As an example, reusing the same `device_resources` instead of creating a new one, specially during search, led to well over a 6x speedup in our recent 4-bit RAFT IVFPQ vs SCANN benchmarks. This matters a lot.